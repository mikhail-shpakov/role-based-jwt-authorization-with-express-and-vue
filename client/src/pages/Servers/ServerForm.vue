<template lang="pug">
.container
  .title.is-size-4 {{ selected.id ? 'Редактировать' : 'Новый сервер' }}
  form.box(
    @submit.prevent="saveChanges"
  )
    template(v-if="selected.id")
      p
        strong ID:
        |  {{ selected.id }}

      .divider

    b-field(label="Название сервера")
      b-input(
        v-model.trim="local.serverName"
        expanded
        required
      )
    b-field.input-margin(label="Тип сервера")
      b-select(
        v-model.trim='local.serverType'
        expanded
      )
        option(
          v-for="option in ['dedicated', 'vds', 'hosting']"
          :key="option"
          :value="option"
        ) {{ option }}

    .buttons.is-pulled-right.button-custom
      b-button(
        v-if="selected.id"
        ref="button-delete"
        type="is-danger"
        icon-left="delete"
        :loading="isLoadingDelete"
        @click.native="confirmDelete()"
      ) Удалить
      b-button(
        ref="button-ed"
        type="is-success"
        :loading="isLoadingChanges"
        :disabled="isCheckStatusButton"
        @click.native="saveChanges(selected.id ? 'edit' : 'add')"
      ) {{ selected.id ? 'Сохранить изменения' : 'Добавить сервер' }}
</template>

<script>
import { mapActions } from 'vuex'

export default {
  name: 'serverForm',
  props: {
    selected: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      isLoadingChanges: false,
      isLoadingDelete: false,
      local: {
        id: '',
        serverName: '',
        serverType: ''
      }
    }
  },
  methods: {
    ...mapActions('server', [
      'EDIT_SERVER',
      'ADD_SERVER',
      'DELETE_SERVER'
    ]),
    async saveChanges (action) {
      if (this.isCheckStatusButton) return

      this.isLoadingChanges = true

      action === 'edit'
        ? await this.EDIT_SERVER(this.local)
        : await this.ADD_SERVER({ serverName: this.local.serverName, serverType: this.local.serverType })

      this.isLoadingChanges = false
    },
    confirmDelete () {
      this.$buefy.dialog.confirm({
        message: `Удалить сервер с названием <strong>${this.local.serverName}</strong>?`,
        cancelText: 'Отмена',
        confirmText: 'Удалить',
        type: 'is-danger',
        onConfirm: async () => { await this.deleteServer() }
      })
    },
    async deleteServer () {
      this.isLoadingDelete = true
      await this.DELETE_SERVER(this.local.id)
      this.isLoadingDelete = false
    }
  },
  computed: {
    isCheckStatusButton () {
      /**
       * Чтобы кнопка сохранения изменений стала активной,
       * поля в форме должны быть не пустыми
       * и отличаться от уже установленных значений
       */
      if (this.local.serverName.length < 3) return true

      const checkDifValue = Object.keys(this.local).some(v =>
        this.local[v] !== this.selected[v]
      )
      return !checkDifValue
    }
  },
  watch: {
    selected: {
      handler () {
        /**
         * Чтобы не менять props напрямую,
         * перезаписываем значения в $data.local
         */
        this.local = Object.assign({}, this.selected)
      },
      immediate: true
    }
  }
}
</script>

<style lang="sass" scoped>
.input-margin
  margin-top: 15px

form.box
  border: 1px solid $color-alt
  border-radius: 8px
  padding: 20px 10px 110px
  box-shadow: none
  @media (min-width: $display-bp-mobile)
    padding: 20px 20px 70px

  .divider
    border-bottom: 1px solid $color-alt
    margin: 15px -10px
    @media (min-width: $display-bp-mobile)
      margin: 15px -20px

  .button-custom
    margin-top: 15px
</style>

<style lang="sass">
input.input
  box-shadow: none
</style>
