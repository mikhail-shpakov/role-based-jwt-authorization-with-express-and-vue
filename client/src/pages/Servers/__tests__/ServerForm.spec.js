import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import Buefy from 'buefy'
import ServerForm from '../ServerForm'

const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(Vuex)

describe('./ServerForm.vue', () => {
  let store
  let actions
  let selected

  beforeEach(() => {
    selected = {
      id: '1',
      serverName: 'Супер сервер',
      serverType: 'dedicated'
    }

    window.confirm = jest.fn(() => true)

    actions = {
      EDIT_SERVER: jest.fn(),
      ADD_SERVER: jest.fn(),
      DELETE_SERVER: jest.fn()
    }

    store = new Vuex.Store({
      modules: {
        server: {
          actions,
          namespaced: true
        }
      }
    })
  })

  it('when load page, ServerForm renders correctly', () => {
    const wrapper = shallowMount(ServerForm, {
      localVue,
      store,
      propsData: { selected }
    })

    expect(wrapper).toMatchSnapshot()
  })

  it('when click button for edit exist server,' +
    'but local input value not changes, will be not called EDIT_SERVER()', () => {
    const wrapper = shallowMount(ServerForm, {
      localVue,
      store,
      propsData: { selected }
    })
    const button = wrapper.find({ ref: 'button-ed' })

    button.trigger('click')

    expect(actions.EDIT_SERVER).toHaveBeenCalled()
  })

  it('when click button for edit exist server,' +
    'and local input value changes, will be called Vuex EDIT_SERVER()', () => {
    const wrapper = shallowMount(ServerForm, {
      localVue,
      store,
      propsData: { selected }
    })
    const button = wrapper.find({ ref: 'button-ed' })
    wrapper.setData({
      local: {
        serverName: 'New name'
      }
    })

    button.trigger('click')

    expect(actions.EDIT_SERVER).toHaveBeenCalled()
  })

  it('when click button for add server,' +
    'and input not empty, will be called Vuex ADD_SERVER()', () => {
    selected.id = 0

    const wrapper = shallowMount(ServerForm, {
      localVue,
      store,
      propsData: { selected }
    })
    const button = wrapper.find({ ref: 'button-ed' })
    wrapper.setData({
      local: {
        serverName: 'New name'
      }
    })

    button.trigger('click')

    expect(actions.ADD_SERVER).toHaveBeenCalled()
  })

  it('when click button for delete server, will be called modal', async () => {
    const wrapper = shallowMount(ServerForm, {
      localVue,
      store,
      propsData: { selected }
    })
    const button = wrapper.find({ ref: 'button-delete' })

    button.trigger('click')

    expect(window.confirm)
  })
})
