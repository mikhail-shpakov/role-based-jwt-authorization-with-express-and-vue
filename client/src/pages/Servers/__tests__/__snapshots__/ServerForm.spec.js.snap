// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./ServerForm.vue when load page, ServerForm renders correctly 1`] = `
<div class="container">
  <div class="title is-size-4">Редактировать</div>
  <form class="box">
    <p><strong>ID:</strong> 1</p>
    <div class="divider"></div>
    <b-field-stub label="Название сервера" addons="true">
      <b-input-stub expanded="true" usehtml5validation="true" value="Супер сервер" type="text" hascounter="true" customclass="" required="required"></b-input-stub>
    </b-field-stub>
    <b-field-stub label="Тип сервера" addons="true" class="input-margin">
      <b-select-stub expanded="true" usehtml5validation="true" value="dedicated">
        <option value="dedicated">dedicated</option>
        <option value="vds">vds</option>
        <option value="hosting">hosting</option>
      </b-select-stub>
    </b-field-stub>
    <div class="buttons is-pulled-right button-custom">
      <b-button-stub type="is-danger" iconleft="delete" nativetype="button" tag="button">Удалить</b-button-stub>
      <b-button-stub type="is-success" nativetype="button" tag="button" disabled="true">Сохранить изменения</b-button-stub>
    </div>
  </form>
</div>
`;

exports[`./ServerForm.vue when load page, ServerForm renders correctly 2`] = `
<div class="container">
  <div class="title is-size-4">Редактировать</div>
  <form class="box">
    <p><strong>ID:</strong> 1</p>
    <div class="divider"></div>
    <b-field-stub label="Название сервера" addons="true">
      <b-input-stub expanded="true" usehtml5validation="true" value="Супер сервер" type="text" hascounter="true" customclass="" required="required"></b-input-stub>
    </b-field-stub>
    <b-field-stub label="Тип сервера" addons="true" class="input-margin">
      <b-select-stub expanded="true" usehtml5validation="true" value="dedicated">
        <option value="dedicated">dedicated</option>
        <option value="vds">vds</option>
        <option value="hosting">hosting</option>
      </b-select-stub>
    </b-field-stub>
    <div class="buttons is-pulled-right button-custom">
      <b-button-stub type="is-danger" iconleft="delete" nativetype="button" tag="button">Удалить</b-button-stub>
      <b-button-stub type="is-success" nativetype="button" tag="button" disabled="true">Сохранить изменения</b-button-stub>
    </div>
  </form>
</div>
`;

exports[`./ServerForm.vue when load page, form renders correctly 1`] = `
<div class="container">
  <div class="title is-size-4">Редактировать</div>
  <form class="box">
    <p><strong>ID:</strong> 1</p>
    <div class="divider"></div>
    <b-field-stub label="Название сервера" addons="true">
      <b-input-stub expanded="true" usehtml5validation="true" value="Супер сервер" type="text" hascounter="true" customclass="" required="required"></b-input-stub>
    </b-field-stub>
    <b-field-stub label="Тип сервера" addons="true" class="input-margin">
      <b-select-stub expanded="true" usehtml5validation="true" value="dedicated">
        <option value="dedicated">dedicated</option>
        <option value="vds">vds</option>
        <option value="hosting">hosting</option>
      </b-select-stub>
    </b-field-stub>
    <div class="buttons is-pulled-right button-custom">
      <b-button-stub type="is-danger" iconleft="delete" nativetype="button" tag="button">Удалить</b-button-stub>
      <b-button-stub type="is-success" nativetype="button" tag="button" disabled="true">Сохранить изменения</b-button-stub>
    </div>
  </form>
</div>
`;
